From 5207a59def35eddb4361124c4586c54834dc213e Mon Sep 17 00:00:00 2001
From: Marcin Naczk <marcin.naczk@intel.com>
Date: Thu, 27 Oct 2022 15:37:27 +0200
Subject: [PATCH] Remove wrong check of __opencl_c_images feature macro

Patched by OpenCL-Clang as WA for clang issue
https://github.com/llvm/llvm-project/issues/58017
Deleted mechanism that wrongly assumes that
all functions that use image types must be under
__opencl_c_images feature macro
---
 clang/lib/Sema/SemaType.cpp | 16 +++++++---------
 1 file changed, 7 insertions(+), 9 deletions(-)

diff --git a/clang/lib/Sema/SemaType.cpp b/clang/lib/Sema/SemaType.cpp
index ab47e9f03eaf..3723f4421057 100644
--- a/clang/lib/Sema/SemaType.cpp
+++ b/clang/lib/Sema/SemaType.cpp
@@ -1734,20 +1734,18 @@ static QualType ConvertDeclSpecToType(TypeProcessingState &state) {
     const auto &OpenCLOptions = S.getOpenCLOptions();
     bool IsOpenCLC30Compatible =
         S.getLangOpts().getOpenCLCompatibleVersion() == 300;
-    // OpenCL C v3.0 s6.3.3 - OpenCL image types require __opencl_c_images
-    // support.
     // OpenCL C v3.0 s6.2.1 - OpenCL 3d image write types requires support
     // for OpenCL C 2.0, or OpenCL C 3.0 or newer and the
     // __opencl_c_3d_image_writes feature. OpenCL C v3.0 API s4.2 - For devices
     // that support OpenCL 3.0, cl_khr_3d_image_writes must be returned when and
     // only when the optional feature is supported
-    if ((Result->isImageType() || Result->isSamplerT()) &&
-        (IsOpenCLC30Compatible &&
-         !OpenCLOptions.isSupported("__opencl_c_images", S.getLangOpts()))) {
-      S.Diag(DS.getTypeSpecTypeLoc(), diag::err_opencl_requires_extension)
-          << 0 << Result << "__opencl_c_images";
-      declarator.setInvalidType();
-    } else if (Result->isOCLImage3dWOType() &&
+    
+    // Patched by OpenCL-Clang as WA for clang issue
+    // https://github.com/llvm/llvm-project/issues/58017
+    // Deleted mechanism that wrongly assumes that
+    // all functions that use image types must be under
+    // __opencl_c_images feature macro
+    if (Result->isOCLImage3dWOType() &&
                !OpenCLOptions.isSupported("cl_khr_3d_image_writes",
                                           S.getLangOpts())) {
       S.Diag(DS.getTypeSpecTypeLoc(), diag::err_opencl_requires_extension)
-- 
2.33.0.windows.1

